### PR TITLE
Cliquet storage should be able to remove tombstones

### PR DESCRIPTION
**Pull request by Natim**
Refs mozilla-services/kinto#136

At the moment, delete_all() remove all objects and create tombstones, we should be able to pass a parameter ``delete_tombstones=True`` to also remove tombstones.

### CODE REVIEW

File: cliquet/storage/__init__.py [hunk 1]
```diff
@@ -199,6 +199,24 @@ def delete_all(self, collection_id, parent_id, filters=None,
         """
         raise NotImplementedError
 
+    def delete_tombstones(self, collection_id, parent_id, before=None,
```
Comment (leplatrem) [inline cliquet/storage/__init__.py]: nit: purge ? purge_deleted ? purge_tombstones ?

Comment (leplatrem) [inline cliquet/storage/__init__.py]: *(distinguish from delete records)*

Comment (Natim) [inline cliquet/storage/__init__.py]: I like ``purge_deleted``

File: cliquet/storage/__init__.py [hunk 2]
```diff
@@ -199,6 +199,24 @@ def delete_all(self, collection_id, parent_id, filters=None,
...
+                          id_field=DEFAULT_ID_FIELD,
+                          modified_field=DEFAULT_MODIFIED_FIELD,
+                          auth=None):
+        """Delete all tombstones for deleted object in this `collection_id`
+        for this `parent_id`.
+
+        :param str collection_id: the collection id.
+        :param str parent_id: the collection parent.
+
+        :param int before: Optionnally elete tombstones older than it.
```
Comment (leplatrem) [inline cliquet/storage/__init__.py]: «Optionnal timestamp to limit deletion (?)»

inclusive ? exclusive ?

Comment (Natim) [inline cliquet/storage/__init__.py]: exclusive, reason why I wanted to use older than.

File: cliquet/storage/postgresql/__init__.py [hunk 3]
```diff
@@ -448,8 +448,48 @@ def delete_all(self, collection_id, parent_id, filters=None,
...
+            DELETE
+            FROM deleted
+            WHERE parent_id = %%(parent_id)s
+              AND collection_id = %%(collection_id)s
+              %(conditions_filter)s
+            RETURNING id
+        )
+        SELECT COUNT(*) as deleted FROM deleted_records
+        RETURNING deleted;
+        """
```
Comment (leplatrem) [inline cliquet/storage/postgresql/__init__.py]: I think you can just DELETE FROM and use cursor.rowcount to get the number of affected rows

Comment (Natim) [inline cliquet/storage/postgresql/__init__.py]: :+1:

**Comment by leplatrem** [comment]
I have one doubt : I am not sure we should mix the delete of records and purge of tombstones.

I don't have a strong opinion about this, but maybe we could have an option that prevents tombstones to be created on deletion, and keep the purge separated.
```python
storage.delete_all(..., tombstones=False)
storage.purge_tombstones(...)
```
Thoughts ?

**Comment by Natim** [comment]
Actually I am not sure we should let people delete things without tombstone. What is the usecase?

**Comment by Natim** [comment]
Also I agree to keep them separated. I did it because when I asked, you told me a param would be nice, apparently we were not talking about the same feature.

**Comment by leplatrem** [comment]
> told me a param would be nice
...handling two channels of communication at the same time

> delete things without tombstone. What is the usecase?
... in Kinto: delete a collection with 12000 records. You don't want the 12000 tombstones to be created if you purge them just after, no ?

**Comment by Natim** [comment]
> You don't want the 12000 tombstones to be created if you purge them just afte

Good point.

**Comment by tarekziade** [comment]
what is a "tombstone" ? I could not find its definition anywhere.

**Comment by leplatrem** [comment]
> what is a "tombstone" ? I could not find its definition anywhere.

You are right. If we introduce this notion in the code, we should add it to the glossary.

So that clients can obtain the list of deleted records when polling for changes, a tombstone is created when to keep track of the deletion.

File: cliquet/storage/__init__.py (hunk 1 above)
Comment (ametaireau) [inline cliquet/storage/__init__.py]: +1

**Comment by Natim** [comment]
Ok so I have updated the code and tests for the purge_deleted part.
I will now start adding the remove without tombstone part as well.

**Pull request merged**